### PR TITLE
Specify headers by environment variable

### PR DESCRIPTION
**Pull request by JacobHenner**
Allow vault CLI HTTP headers to be specified by environment variable. This is especially useful in situations where header values should not be exposed as `vault` command parameters (secrets, for example). This builds on the work introduced in https://github.com/hashicorp/vault/issues/8754.

Fixes https://github.com/hashicorp/vault/issues/16663

### CODE REVIEW

File: api/client.go [hunk 1]
```diff
@@ -661,6 +663,25 @@ func NewClient(c *Config) (*Client, error) {
 		client.setNamespace(namespace)
 	}
 
+	if envHeaders := os.Getenv(EnvVaultHeaders); envHeaders != "" {
+		var result map[string]any
+		err := json.Unmarshal([]byte(envHeaders), &result)
```
Comment (JacobHenner) [inline api/client.go:668]: Parsing JSON objects is the most easily-understood way to support > 1 header specified in the same environment variable. However, it limits each key to a single value (which is not a restriction of HTTP). Is this sufficient, or should an alternative be considered?

Comment (VioletHynes) [reply api/client.go:672]: To me, this seems reasonable. It's imperfect, but it follows the same pattern as is done with the `-header` option, as far as I can tell: https://github.com/hashicorp/vault/pull/12508

File: api/client.go [hunk 2]
```diff
@@ -665,6 +667,25 @@ func NewClient(c *Config) (*Client, error) {
...
+		if err != nil {
+			return nil, fmt.Errorf("could not unmarshal environment-supplied headers")
+		}
+		var forbiddenHeaders []string
+		for key, value := range result {
+			if strings.HasPrefix(key, "X-Vault-") {
+				forbiddenHeaders = append(forbiddenHeaders, key)
+				continue
+			}
+			client.AddHeader(key, value.(string))
```
Comment (VioletHynes) [inline api/client.go:682]: We should add a check for if the value is not a string, and error if it's not, since it could be `any`

**Review by VioletHynes — COMMENTED** [review]
Hey there @JacobHenner ! Sorry that this has lingered for a while, but this looks great.

As well as the comment I left about the conversion, there's a few other things I'd like to see before we approve and merge this:
- The name of the environment variable should be in the changelog!
- An entry in `website/content/docs/commands/index.mdx` (our docs for the CLI) for the header and what it does (plus an example!)
- A test of the behaviour, to make sure it behaves as expected. For prior art, https://github.com/hashicorp/vault/pull/12508 might be a good starting point

If all of that is addressed, I'd be happy to get this merged. Let me know if you have any questions, and thank you for your patience in getting eyes on this.

**Comment by JacobHenner** [comment]
> Hey there @JacobHenner ! Sorry that this has lingered for a while, but this looks great.
> 
> As well as the comment I left about the conversion, there's a few other things I'd like to see before we approve and merge this:
> 
>     * The name of the environment variable should be in the changelog!
> 
>     * An entry in `website/content/docs/commands/index.mdx` (our docs for the CLI) for the header and what it does (plus an example!)
> 
>     * A test of the behaviour, to make sure it behaves as expected. For prior art, [CLI: add new -header option to be able to add headers to all cli http request #12508](https://github.com/hashicorp/vault/pull/12508) might be a good starting point
> 
> 
> If all of that is addressed, I'd be happy to get this merged. Let me know if you have any questions, and thank you for your patience in getting eyes on this.

Done

**Comment by VioletHynes** [comment]
@JacobHenner One last thing before I can merge this -- this is failing our linter, could you run `make fmt` and push that commit? Once that's added, I can merge this! Everything else is passing!

Thanks in advance :)

**Comment by JacobHenner** [comment]
> @JacobHenner One last thing before I can merge this -- this is failing our linter, could you run `make fmt` and push that commit? Once that's added, I can merge this! Everything else is passing!

Done, sorry for not catching earlier.

**Comment by VioletHynes** [comment]
Merged! Thanks again for this contribution, and we really appreciate your patience in getting this seen to and merged. 

🚢 🚢 🚢

**Pull request merged**